File: 프로그래머스/1/133499. 옹알이 （2）/옹알이 （2）.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

bool isValid(string bab, unordered_set<string>& validWords){
    string prev = "";
    string cur = "";
    
    for(auto c : bab){
        cur += c;
        if(validWords.find(cur) != validWords.end() && cur != prev){
            prev = cur;
            cur = "";
        }
    }
    
    return cur.empty();
}

int solution(vector<string> babbling) {
    int answer = 0;
    unordered_set<string> validWords = {"aya", "ye", "woo", "ma"};
    for(auto bab : babbling)
        if(isValid(bab, validWords))
            answer++;
    return answer;
}

File: 프로그래머스/1/133499. 옹알이 （2）/옹알이 （2）.java
class Solution {
    public int solution(String[] babbling) {
        String[] possibleSounds = {"aya", "ye", "woo", "ma"};
        int count = 0;
        
        for (String word : babbling) {
            if (canPronounce(word, possibleSounds)) {
                count++;
            }
        }
        
        return count;
    }
    
    private boolean canPronounce(String word, String[] possibleSounds) {
        StringBuilder sb = new StringBuilder(word);
        String lastSound = "";
        
        while (sb.length() > 0) {
            boolean found = false;
            for (String sound : possibleSounds) {
                if (sb.indexOf(sound) == 0 && !sound.equals(lastSound)) {
                    sb.delete(0, sound.length());
                    lastSound = sound;
                    found = true;
                    break;
                }
            }
            if (!found) {
                return false;
            }
        }
        
        return true;
    }
}